lib/hashtable.cc

```diff
@@ -1079,5 +1079,79 @@ const
     return size;
 }
 
-// vim: set sts=2 sw=2:
+// Starting from the given seed k-mer, assemble the maximal linear path in
+// both directions.
+//
+// No guarantees on direction, of course - this may return the reverse
+// complement of the input sequence.
+//
+// Note: as written, will ignore branches to the left and continue
+// past them; this probably needs to be fixed.  For now, this means
+// that assembling from two different directions may yield different
+// results.
+
+std::string Hashtable::assemble_linear_path(const Kmer seed_kmer,
+                                            const Hashtable * stop_bf)
+    const
+{
+    std::string start_kmer = seed_kmer.get_string_rep(_ksize);
+    std::string right = _assemble_right(start_kmer.c_str(), stop_bf);
+
+    start_kmer = _revcomp(start_kmer);
+    std::string left = _assemble_right(start_kmer.c_str(), stop_bf);
+
+    left = left.substr(_ksize);
+    return _revcomp(left) + right;
+}
+
+std::string Hashtable::_assemble_right(const char * start_kmer,
+                                       const Hashtable * stop_bf)
+    const
+{
+    const char bases[] = "ACGT";
+    std::string kmer = start_kmer;
+    std::string contig = kmer;
+
+    // This loop extends the starting k-mer to the right as long as it can
+    // do so unambiguously (or not at all).  This involves checking each
+    // possible nucleotide suffix for presence; extension is continued until
+    // either more than one such k-mer is present ('found2' is true), or no
+    // such k-mer is present ('found' is false).
+
+    while (1) {
+        const char * base = &bases[0];
+        bool found = false;
+        char found_base;
+        bool found2 = false;
+
+        // check all four suffixes for presence.
+        while(*base != 0) {
+            std::string try_kmer = kmer.substr(1) + (char) *base;
+
+            // a hit!
+            if (this->get_count(try_kmer.c_str()) &&
+                (!stop_bf || !stop_bf->get_count(try_kmer.c_str()))) {
+                if (found) {
+                    found2 = true;
+                    break;
+                }
+                found_base = (char) *base;
+                found = true;
+            }
+            base++;
+        }
+
+        // exit condition: no suffix k-mer, or more than one.
+        if (!found or found2) {
+            break;
+        }
 
+        // extend assembly!
+        contig += found_base;
+        kmer = kmer.substr(1) + found_base;
+        found = true;
+    }
+    return contig;
+}
```

Comment:
The pointer arithmetic is bit old school :-), but the implementation is clear regardless.